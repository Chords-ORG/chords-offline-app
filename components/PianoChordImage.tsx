import React from "react";
import { StyleSheet, View, Image, Text } from "react-native";

export interface PianoChordImageProps {
  fingers: string[];
}

export default function PianoChordImage({ fingers }: PianoChordImageProps) {
  const pianoBack = require("../assets/images/piano_chord/piano_back.png");
  const blackFinger = require("../assets/images/piano_chord/black_finger.png");
  const whiteFinger = require("../assets/images/piano_chord/white_finger.png");

  const whiteFingersHouses = [2, 4, 7, 9, 11, 14, 16, 19, 21, 23, 26, 28];
  return (
    <View style={styles.container}>
      <Image source={pianoBack} style={styles.pianoBack} />
      {fingers.map((_house: string, i: number) => {
        const house = parseInt(_house);
        return (
          <Image
            key={i}
            source={
              whiteFingersHouses.includes(house) ? whiteFinger : blackFinger
            }
            style={fingerStyles[house]}
          />
        );
      })}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    height: 50,
    width: 200,
  },
  message: {
    fontFamily: "roboto-bold",
    fontSize: 12,
    textAlign: "center",
  },
  pianoBack: {
    height: 50,
    width: 200,
  },
  finger0: {
    position: "absolute",
    height: 8,
    width: 8,
  },
  finger1: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 2,
  },
  finger2: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 8,
  },
  finger3: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 14,
  },
  finger4: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 20,
  },
  finger5: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 26,
  },
  finger6: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 37,
  },
  finger7: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 43,
  },
  finger8: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 49,
  },
  finger9: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 55,
  },
  finger10: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 61,
  },
  finger11: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 67,
  },
  finger12: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 73,
  },
  finger13: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 85,
  },
  finger14: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 90,
  },
  finger15: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 96,
  },
  finger16: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 102,
  },
  finger17: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 108,
  },
  finger18: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 120,
  },
  finger19: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 125,
  },
  finger20: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 132,
  },
  finger21: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 137,
  },
  finger22: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 143,
  },
  finger23: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 149,
  },
  finger24: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 155,
  },
  finger25: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 167,
  },
  finger26: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 173,
  },
  finger27: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 179,
  },
  finger28: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 20,
    left: 185,
  },
  finger29: {
    position: "absolute",
    height: 8,
    width: 8,
    bottom: 2,
    left: 190,
  },
  finger30: {
    position: "absolute",
    height: 0,
    width: 0,
  },
});

const fingerStyles = [
  styles.finger0,
  styles.finger1,
  styles.finger2,
  styles.finger3,
  styles.finger4,
  styles.finger5,
  styles.finger6,
  styles.finger7,
  styles.finger8,
  styles.finger9,
  styles.finger10,
  styles.finger11,
  styles.finger12,
  styles.finger13,
  styles.finger14,
  styles.finger15,
  styles.finger16,
  styles.finger17,
  styles.finger18,
  styles.finger19,
  styles.finger20,
  styles.finger21,
  styles.finger22,
  styles.finger23,
  styles.finger24,
  styles.finger25,
  styles.finger26,
  styles.finger27,
  styles.finger28,
  styles.finger29,
  styles.finger30,
];
