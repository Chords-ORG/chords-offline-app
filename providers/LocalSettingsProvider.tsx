import React from "react";
import { LightStyle, DarkStyle } from "../constants/Styles";
import useLocalConfiguration from "../hooks/useLocalConfiguration";
import { useColorScheme } from "react-native";
import { getItem } from "../functions/storage";

export type ChordType = "sharp" | "bemol";
export type Instrument = "guitar" | "piano";
export type CapoConfig = "auto" | "never";
export type LocalColorScheme = "system" | "light" | "dark";

export const CHORD_TYPE_KEY = "chordType";
export const INSTRUMENT_KEY = "instrument";
export const CAPO_CONFIG_KEY = "capoConfig";
export const LOCAL_COLOR_SCHEME_KEY = "localColorScheme";

interface LocalSettingsContextProps {
  chordType: ChordType;
  instrument: Instrument;
  capoConfig: CapoConfig;
  localColorScheme: LocalColorScheme;
  toggleLocalSettings: () => Promise<void>;
}

export const LocalSettingsContext =
  React.createContext<LocalSettingsContextProps>({
    chordType: "sharp",
    instrument: "guitar",
    capoConfig: "auto",
    localColorScheme: "system",
    toggleLocalSettings: async () => {},
  });

interface LocalSettingsProviderProps {
  children: React.ReactNode;
}

export const LocalSettingsProvider = ({
  children,
}: LocalSettingsProviderProps) => {
  const [chordType, setChordTypeState] = React.useState<ChordType>("sharp");
  const [instrument, setInstrumentState] = React.useState<Instrument>("guitar");
  const [capoConfig, setCapoConfigState] = React.useState<CapoConfig>("auto");
  const [localColorScheme, setLocalColorSchemeState] =
    React.useState<LocalColorScheme>("system");

  const getLocalConfiguration = async () => {
    const chordType = await getItem(CHORD_TYPE_KEY);
    const instrument = await getItem(INSTRUMENT_KEY);
    const capoConfig = await getItem(CAPO_CONFIG_KEY);
    const localColorScheme = await getItem(LOCAL_COLOR_SCHEME_KEY);
    return { chordType, instrument, capoConfig, localColorScheme };
  };

  const toggleLocalSettings = React.useCallback(async () => {
    const localConfiguration = await getLocalConfiguration();
    if (localConfiguration.chordType !== null) {
      setChordTypeState(localConfiguration.chordType as ChordType);
    }
    if (localConfiguration.instrument !== null) {
      setInstrumentState(localConfiguration.instrument as Instrument);
    }
    if (localConfiguration.capoConfig !== null) {
      setCapoConfigState(localConfiguration.capoConfig as CapoConfig);
    }
    if (localConfiguration.localColorScheme !== null) {
      setLocalColorSchemeState(
        localConfiguration.localColorScheme as LocalColorScheme
      );
    }
  }, [
    setChordTypeState,
    setInstrumentState,
    setCapoConfigState,
    setLocalColorSchemeState,
  ]);

  React.useEffect(() => {
    toggleLocalSettings();
  }, []);

  return (
    <LocalSettingsContext.Provider
      value={{
        chordType,
        instrument,
        capoConfig,
        localColorScheme,
        toggleLocalSettings,
      }}
    >
      {children}
    </LocalSettingsContext.Provider>
  );
};
