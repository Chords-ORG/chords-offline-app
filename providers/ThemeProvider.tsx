import React from "react";
import { LightStyle, DarkStyle } from "../constants/Styles";
import useLocalConfiguration from "../hooks/useLocalConfiguration";
import { useColorScheme } from "react-native";
import { getItem } from "../functions/storage";

interface ThemeContextProps {
  styleSheet: typeof LightStyle | typeof DarkStyle;
  theme: "light" | "dark";
  toggleTheme: () => void;
}

export const ThemeContext = React.createContext<ThemeContextProps>({
  styleSheet: LightStyle,
  theme: "light",
  toggleTheme: () => {},
});

interface ThemeProviderProps {
  children: React.ReactNode;
}
export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const [theme, setTheme] = React.useState<"light" | "dark">("light");
  const systemColorScheme = useColorScheme();
  const { localColorScheme } = useLocalConfiguration();

  const toggleTheme = async () => {
    if (localColorScheme === "system") {
      setTheme(systemColorScheme || "light");
    } else {
      setTheme(localColorScheme as "light" | "dark");
    }
  };

  React.useEffect(() => {
    toggleTheme();
  }, [systemColorScheme, localColorScheme]);

  return (
    <ThemeContext.Provider
      value={{
        styleSheet: theme === "light" ? LightStyle : DarkStyle,
        theme,
        toggleTheme,
      }}
    >
      {children}
    </ThemeContext.Provider>
  );
};
