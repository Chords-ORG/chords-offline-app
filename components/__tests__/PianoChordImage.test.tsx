import renderer from "react-test-renderer";
import React from "react";
import { ThemeContext } from "../../providers/ThemeProvider";
import {
  DarkColors,
  DarkStyle,
  LightColors,
  LightStyle,
} from "../../constants/Styles";
import PianoChordImage from "../PianoChordImage";

jest.mock("@react-native-async-storage/async-storage", () =>
  require("@react-native-async-storage/async-storage/jest/async-storage-mock")
);

describe("<PianoChordImage/>", () => {
  describe("Testing sample chords positions on database", () => {
    const pianoChords: Record<
      string,
      string[][]
    > = require("../../assets/data/piano_chords.json");

    const sample = Object.keys(pianoChords).slice(0, 5);

    sample.forEach((chordName) => {
      const fingersPositionsList = pianoChords[chordName];
      fingersPositionsList.forEach((fingers, idx) => {
        it(`should render ${chordName} on position ${idx} on light theme`, () => {
          const tree = renderer
            .create(
              <ThemeContext.Provider
                value={{
                  styleSheet: LightStyle,
                  colors: LightColors,
                  theme: "light",
                }}
              >
                <PianoChordImage fingers={fingers} />
              </ThemeContext.Provider>
            )
            .toJSON();
          expect(tree).toMatchSnapshot();
        });

        it(`should render ${chordName} on position ${idx} on dark theme`, () => {
          const tree = renderer
            .create(
              <ThemeContext.Provider
                value={{
                  styleSheet: DarkStyle,
                  colors: DarkColors,
                  theme: "dark",
                }}
              >
                <PianoChordImage fingers={fingers} />
              </ThemeContext.Provider>
            )
            .toJSON();
          expect(tree).toMatchSnapshot();
        });
      });
    });
  });
});
