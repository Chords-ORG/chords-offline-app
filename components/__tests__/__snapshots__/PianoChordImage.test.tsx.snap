// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PianoChordImage/> Testing sample chords positions on database should render C9 on position 0 on dark theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 2,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 26,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 49,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render C9 on position 0 on light theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 2,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 26,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 49,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render Dm7 on position 0 on dark theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 37,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 85,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render Dm7 on position 0 on light theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 37,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 85,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render Dm7 on position 1 on dark theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 37,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 85,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render Dm7 on position 1 on light theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 37,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 85,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 14,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render F/A on position 0 on dark theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 120,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 85,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render F/A on position 0 on light theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 120,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 61,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 85,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render G on position 0 on dark theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 49,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 73,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 96,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render G on position 0 on light theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 49,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 73,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 96,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render G/B on position 0 on dark theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 73,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 132,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 73,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 96,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`<PianoChordImage/> Testing sample chords positions on database should render G/B on position 0 on light theme 1`] = `
<View
  style={
    {
      "height": 50,
      "width": 200,
    }
  }
>
  <Image
    source={1}
    style={
      {
        "height": 50,
        "width": 200,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 73,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 132,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 73,
        "position": "absolute",
        "width": 8,
      }
    }
  />
  <Image
    source={1}
    style={
      {
        "bottom": 2,
        "height": 8,
        "left": 96,
        "position": "absolute",
        "width": 8,
      }
    }
  />
</View>
`;
